util/topgen/merge.py

```diff
@@ -549,20 +558,24 @@ def extract_clocks(top: OrderedDict):
         ep_name = ep['name']
         ep_clks = []
 
-        group = clocks.groups[ep_grp]
-
         for port, clk in ep['clock_srcs'].items():
+
+            # The value of a particular connection is a dict,
+            # there are additional attributes to explore
+            if isinstance(clk, Dict):
+                if 'group' in clk:
+                    group = clocks.groups[clk['group']]
+                clk = clk['clock']
+            else:
+                group = clocks.groups[ep_grp]
```

Comment:
This is neat, but it might be a bit easier to read if you did it "the other way around". Something like this, maybe?
```python
    if isinstance(clk, str):
        group_name = None
        clk_name = clk
    else:
        assert isinstance(clk, dict)
        group_name = clk.get('group')
        clk_name = clk['clock']

    group = clocks.groups[group_name or ep_grp]
```
(untested!)

This also handles the case where `clk` is a dictionary but doesn't have a `group` key, which I think is missing in the existing code.

Reply:
yep updated! i also massaged the names slightly to make it less confusing.

Reply:
Oh! That's much neater than what I was suggesting. Nice!